src/main/java/fopbot/KarelWorld.java

```diff
@@ -118,19 +122,32 @@ public KarelWorld(int width, int height) {
     }
 
     /**
-     * Adds the specified robot to this world.
+     * Adds the specified {@link Robot} to this world.
      *
      * @param robot the robot to place
      */
     public void addRobot(Robot robot) {
         fields[robot.getY()][robot.getX()].getEntities().add(robot);
-        robot.setId(Integer.toString(robotCount));
+        robot.setId(Integer.toString(nextRobotId));
+        nextRobotId++;
         robotCount++;
         traces.put(robot.getId(), new RobotTrace());
         triggerUpdate();
         sleep();
     }
 
+    /**
+     * Removes the specified {@link Robot} from this world.
+     *
+     * @param robot the robot to remove
+     */
```

Comment:
Maybe we should add a note that this will in fact not free any memory since the robot instance is still accessible through the trails